File: job-2022-autumn/1081.smallest-subsequence-of-distinct-characters.js
/*
 * @lc app=leetcode id=1081 lang=javascript
 *
 * [1081] Smallest Subsequence of Distinct Characters
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
var smallestSubsequence = function (s) {
    /** from bottom to top, the letter increases.
     * @type {string[]}
     */
    const monostack = [];
    const letterToLeftCount = {};
    const inStackLetters = new Set();

    for (const c of s) {
        letterToLeftCount[c] = (letterToLeftCount[c] ?? 0) + 1;
    }

    for (const c of s) {
        if (monostack.length === 0) {
            monostack.push(c);
            inStackLetters.add(c);
        } else if (!inStackLetters.has(c)) {
            while (
                monostack.length > 0 &&
                monostack[monostack.length - 1] >= c &&
                letterToLeftCount[monostack[monostack.length - 1]] > 0
            ) {
                inStackLetters.delete(monostack.pop());
            }
            monostack.push(c);
            inStackLetters.add(c);
        }
        letterToLeftCount[c]--;
    }

    return monostack.join('');
};
// @lc code=end

File: src/2026/1081.smallest-subsequence-of-distinct-characters.cpp
/*
 * @lc app=leetcode id=1081 lang=cpp
 *
 * [1081] Smallest Subsequence of Distinct Characters
 */

#include <algorithm>
#include <array>
#include <stack>
#include <string>

// @lc code=start
class Solution {
 public:
  std::string smallestSubsequence(std::string s) {
    std::array<int, 26> remaining_char_freqs = {};
    for (const char c : s) {
      remaining_char_freqs[c - 'a']++;
    }
    std::array<bool, 26> is_in_stack = {};

    std::stack<char> increasing_stack;

    for (const char c : s) {
      if (is_in_stack[c - 'a']) {
        remaining_char_freqs[c - 'a']--;
        continue;
      }
      is_in_stack[c - 'a'] = true;

      while (!increasing_stack.empty() && increasing_stack.top() >= c &&
             // We can not remove the top char if non left in s.
             remaining_char_freqs[increasing_stack.top() - 'a'] > 0) {
        is_in_stack[increasing_stack.top() - 'a'] = false;
        increasing_stack.pop();
      }

      increasing_stack.push(c);
      remaining_char_freqs[c - 'a']--;
    }

    std::string result;
    result.reserve(increasing_stack.size());
    while (!increasing_stack.empty()) {
      result.push_back(increasing_stack.top());
      increasing_stack.pop();
    }

    std::ranges::reverse(result);
    return result;
  }
};
// @lc code=end

int main() {
  Solution sol;
  sol.smallestSubsequence("bcbcbcababa");
}
